### PR TITLE
Martin/ios 559 read animation

### PR DESCRIPTION
**Pull request by MartinJibber**
This also fixes some image loading flickering issues.

After experimenting a bit. I opted to go for animating the alpha on images.

I tried blurring at first, but it didn't always work because blur also means the image is loading. We can revisit later if it still doesn't feel right, but I actually think it looks pretty nice when you also have text.

https://user-images.githubusercontent.com/88675954/164117936-374a7d45-0090-44c2-aa50-9471196d3518.mov

### CODE REVIEW

File: Shared/UI/Views/DisplayableImageView.swift [hunk 1]
```diff
@@ -14,11 +14,6 @@ import Lottie
 
 class DisplayableImageView: BaseView {
 
-    private(set) var imageView = UIImageView()
-    var cancellables = Set<AnyCancellable>()
-
-    let blurView = BlurView()
-
```
Comment (MartinJibber) [inline Shared/UI/Views/DisplayableImageView.swift:21]: Grouping the views together.

File: Shared/UI/Views/DisplayableImageView.swift [hunk 2]
```diff
@@ -28,19 +23,27 @@ class DisplayableImageView: BaseView {
...
             // A nil displayable can be applied immediately without creating a task.
             guard let displayableRef = self.displayable else {
                 self.imageView.image = nil
                 return
             }
 
+            // Don't load the displayable again if it hasn't changed.
+            if let displayable = self.displayable, displayable.isEqual(to: oldValue) {
+                return
+            }
```
Comment (MartinJibber) [inline Shared/UI/Views/DisplayableImageView.swift:38]: Got the optimization working properly finally.

File: Shared/UI/Views/DisplayableImageView.swift [hunk 3]
```diff
@@ -28,16 +23,26 @@ class DisplayableImageView: BaseView {
...
+
             self.displayableTask?.cancel()
 
             // A nil displayable can be applied immediately without creating a task.
             guard let displayableRef = self.displayable else {
                 self.imageView.image = nil
+                self.state = .initial
```
Comment (MartinJibber) [inline Shared/UI/Views/DisplayableImageView.swift:45]: So things don't get stuck in a load state.

File: Shared/UI/Views/MessageContentView/MessageContentView.swift [hunk 4]
```diff
@@ -265,9 +268,14 @@ class MessageContentView: BaseView {
 
             switch message.kind {
             case .photo(photo: let photo, _):
+                // Only reload the picture if it's actually a new message.
```
Comment (MartinJibber) [inline Shared/UI/Views/MessageContentView/MessageContentView.swift:271]: This fixes issues with the photos flickering the blur.

File: Shared/UI/Views/MessageContentView/MessageContentView.swift [hunk 5]
```diff
@@ -318,6 +326,14 @@ class MessageContentView: BaseView {
         self.bubbleView.orientation = tailOrientation
     }
 
+    func playReadAnimations() async {
+        await self.textView.startReadAnimation()
+        await UIView.awaitAnimation(with: .custom(1)) {
+            self.imageView.alpha = 1
+            self.linkView.alpha = 1
+        }
```
Comment (MartinJibber) [inline Shared/UI/Views/MessageContentView/MessageContentView.swift:334]: First we play the text, then the image animations.

File: iOS/Models/Attachments/PhotoAttachment.swift [hunk 6]
```diff
@@ -34,9 +31,28 @@ struct PhotoAttachment: MediaItem {
...
-    var _data: Data?
     var info: [AnyHashable: Any]?
+
+    init(url: URL?, previewUrl: URL?, data: Data?, info: [AnyHashable : Any]?) {
+        self.url = url
+        self.previewUrl = previewUrl
+        self.data = data
+        self.info  = info
+
+        self.convertDataIntoImage()
```
Comment (MartinJibber) [inline iOS/Models/Attachments/PhotoAttachment.swift:47]: The photo attachment would make a new image every time the image was queried. Each image was a different instance so the displayable image view thought they were different.

This helps fix that issue so we don't load the same image multiple times.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**